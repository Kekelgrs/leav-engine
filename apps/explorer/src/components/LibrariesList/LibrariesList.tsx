// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {useMutation, useQuery} from '@apollo/client';
import {Divider, PageHeader, Row, Spin} from 'antd';
import {saveUserData} from 'graphQL/mutations/userData/saveUserData';
import {getUserDataQuery} from 'graphQL/queries/userData/getUserData';
import React, {useEffect, useState} from 'react';
import {useTranslation} from 'react-i18next';
import {useParams} from 'react-router-dom';
import {setNotificationBase} from 'redux/notifications';
import {useAppDispatch} from 'redux/store';
import {getLibrariesListQuery} from '../../graphQL/queries/libraries/getLibrariesListQuery';
import {SAVE_USER_DATA, SAVE_USER_DATAVariables} from '../../_gqlTypes/SAVE_USER_DATA';
import {IBaseNotification, NotificationType} from '../../_types/types';
import ErrorDisplay from '../shared/ErrorDisplay';
import LibraryCard from './LibraryCard';
import LibraryDetail from './LibraryDetail';

export const FAVORITE_LIBRARIES_KEY = 'favorites_libraries_ids';

function LibrariesList(): JSX.Element {
    const {t} = useTranslation();
    const {libId, libQueryName, filterName} = useParams<{libId: string; libQueryName: string; filterName: string}>();

    const dispatch = useAppDispatch();

    const [activeLibrary, setActiveLibrary] = useState<string>(libId);

    const librariesListQuery = useQuery(getLibrariesListQuery);
    const userDataQuery = useQuery(getUserDataQuery, {
        variables: {key: FAVORITE_LIBRARIES_KEY}
    });

    const [updateFavoritesMutation] = useMutation<SAVE_USER_DATA, SAVE_USER_DATAVariables>(saveUserData, {
        refetchQueries: [
            {
                query: getUserDataQuery,
                variables: {key: FAVORITE_LIBRARIES_KEY}
            }
        ]
    });

    const onUpdateFavorite = async (id: string) => {
        await updateFavoritesMutation({
            variables: {
                key: FAVORITE_LIBRARIES_KEY,
                value: favoriteIds.includes(id) ? favoriteIds.filter(e => e !== id) : favoriteIds.concat([id]),
                global: false
            }
        });
    };

    useEffect(() => {
        const baseNotification: IBaseNotification = {
            content: t('notification.base-message'),
            type: NotificationType.basic
        };
        dispatch(setNotificationBase(baseNotification));
    }, [t, dispatch]);

    useEffect(() => {
        setActiveLibrary(libId);
    }, [libId, setActiveLibrary]);

    if (librariesListQuery.loading || userDataQuery.loading) {
        return <Spin />;
    }

    if (librariesListQuery.error || userDataQuery.error) {
        return <ErrorDisplay message={librariesListQuery.error?.message || userDataQuery.error?.message} />;
    }

    const libraries = librariesListQuery.data?.libraries?.list ?? [];
    const favoriteIds = userDataQuery.data?.userData?.data ?? [];

    return (
        <div className="wrapper-page">
            <PageHeader title={t('lib_list.header')} />
            <Row gutter={[16, 16]}>
                {libraries
                    .filter(lib => favoriteIds.includes(lib.id))
                    .map(lib => (
                        <LibraryCard
                            active={lib.id === activeLibrary}
                            key={lib.id}
                            lib={lib}
                            isFavorite={true}
                            onUpdateFavorite={onUpdateFavorite}
                        />
                    ))}
            </Row>
            {favoriteIds.length > 0 && <Divider />}
            <Row gutter={[16, 16]}>
                {libraries
                    .filter(lib => !favoriteIds.includes(lib.id))
                    .map(lib => (
                        <LibraryCard
                            active={lib.id === activeLibrary}
                            key={lib.id}
                            lib={lib}
                            onUpdateFavorite={onUpdateFavorite}
                        />
                    ))}
            </Row>
            {libId && libQueryName && (
                <>
                    <Divider />
                    <LibraryDetail libId={libId} libQueryName={libQueryName} filterName={filterName} />
                </>
            )}
        </div>
    );
}

export default LibrariesList;
