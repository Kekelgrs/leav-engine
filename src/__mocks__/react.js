const react = require('react');
module.exports = {...react, memo: x => x};

export default undefined;
