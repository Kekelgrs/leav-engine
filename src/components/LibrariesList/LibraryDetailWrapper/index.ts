import LibraryDetailWrapper from './LibraryDetailWrapper';
export default LibraryDetailWrapper;
