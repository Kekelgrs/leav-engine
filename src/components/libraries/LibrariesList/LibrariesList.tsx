import React from 'react';
import {withNamespaces, WithNamespaces} from 'react-i18next';
import {Checkbox, Icon, Input, Table} from 'semantic-ui-react';
import {localizedLabel} from '../../../utils/utils';
import {GET_LIBRARIES_libraries} from '../../../_gqlTypes/GET_LIBRARIES';
import Loading from '../../shared/Loading';
import DeleteLibrary from '../DeleteLibrary';

interface ILibrariesListProps extends WithNamespaces {
    libraries: GET_LIBRARIES_libraries[] | null;
    onRowClick: (library: GET_LIBRARIES_libraries) => void;
    onFiltersUpdate?: (filters: any) => void;
    loading?: boolean;
    filters?: any;
}
function LibrariesList({
    libraries,
    onRowClick,
    onFiltersUpdate,
    loading,
    filters,
    t,
    i18n: i18next
}: ILibrariesListProps): JSX.Element {
    const _handleFilterChange = (e: React.SyntheticEvent, d: any) => {
        // If a checkbox was not checked and is clicked, go back to indeterminate state
        if (d.type === 'checkbox' && filters.system === false && d.checked) {
            d.indeterminate = true;
            d.checked = false;
        }

        if (onFiltersUpdate) {
            onFiltersUpdate(d);
        }
    };

    return (
        <React.Fragment>
            <Table selectable striped>
                <Table.Header>
                    <Table.Row>
                        <Table.HeaderCell width={1} />
                        <Table.HeaderCell width={7}>{t('libraries.label')}</Table.HeaderCell>
                        <Table.HeaderCell width={6}>{t('libraries.ID')}</Table.HeaderCell>
                        <Table.HeaderCell width={1}>{t('libraries.isSystem')}</Table.HeaderCell>
                        <Table.HeaderCell width={1} />
                    </Table.Row>
                    <Table.Row className="filters">
                        <Table.HeaderCell />
                        <Table.HeaderCell>
                            <Input
                                size="small"
                                fluid
                                placeholder={t('libraries.label') + '...'}
                                name="label"
                                value={filters.label || ''}
                                onChange={_handleFilterChange}
                            />
                        </Table.HeaderCell>
                        <Table.HeaderCell>
                            <Input
                                size="small"
                                fluid
                                placeholder={t('libraries.ID') + '...'}
                                name="id"
                                value={filters.id || ''}
                                onChange={_handleFilterChange}
                            />
                        </Table.HeaderCell>
                        <Table.HeaderCell>
                            <Checkbox
                                indeterminate={typeof filters.system === 'undefined'}
                                name="system"
                                checked={filters.system}
                                onChange={_handleFilterChange}
                            />
                        </Table.HeaderCell>
                        <Table.HeaderCell />
                    </Table.Row>
                </Table.Header>
                <Table.Body>
                    {loading ? (
                        <Table.Row>
                            <Table.Cell colSpan={6}>
                                <Loading />
                            </Table.Cell>
                        </Table.Row>
                    ) : (
                        libraries &&
                        libraries.map(l => {
                            const libLabel = localizedLabel(l.label, i18next);
                            const onClick = () => onRowClick(l);
                            return (
                                <Table.Row key={l.id} onClick={onClick}>
                                    <Table.Cell>
                                        <Icon name="book" size="large" />
                                    </Table.Cell>
                                    <Table.Cell>{libLabel}</Table.Cell>
                                    <Table.Cell>{l.id}</Table.Cell>
                                    <Table.Cell>
                                        <Checkbox readOnly checked={!!l.system} />
                                    </Table.Cell>
                                    <Table.Cell>
                                        <DeleteLibrary library={l} />
                                    </Table.Cell>
                                </Table.Row>
                            );
                        })
                    )}
                </Table.Body>
            </Table>
        </React.Fragment>
    );
}

LibrariesList.defaultProps = {
    loading: false,
    libraries: [],
    filters: {}
};

export default withNamespaces()(LibrariesList);
