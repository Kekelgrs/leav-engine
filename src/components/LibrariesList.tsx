import * as React from 'react';
import {Link} from 'react-router-dom';
import {List} from 'semantic-ui-react';
import {GET_LIBRARIES_libraries} from '../_types/GET_LIBRARIES';

interface ILibrariesListProps {
    libraries: Array<GET_LIBRARIES_libraries | null> | null;
}

function LibrariesList({libraries}: ILibrariesListProps): JSX.Element {
    return (
        <List divided relaxed>
            {libraries === null ? (
                <p>No Libraries</p>
            ) : (
                libraries.map((l: GET_LIBRARIES_libraries) => {
                    const label = l.label !== null ? l.label.fr || l.label.en : l.id;
                    return (
                        <List.Item as={Link} to="/libraries" key={l.id}>
                            <List.Icon name="book" size="large" />
                            <List.Content>
                                <List.Header size="huge">{label}</List.Header>
                                <List.Description>{l.id}</List.Description>
                            </List.Content>
                        </List.Item>
                    );
                })
            )}
        </List>
    );
}

export default LibrariesList;
