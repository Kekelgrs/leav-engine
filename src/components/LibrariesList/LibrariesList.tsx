import {i18n, TranslationFunction} from 'i18next';
import * as React from 'react';
import {translate} from 'react-i18next';
import {Checkbox, Icon, Table} from 'semantic-ui-react';
import DeleteLibrary from 'src/containers/DeleteLibrary';
import {localizedLabel} from 'src/utils/utils';
import {GET_LIBRARIES_libraries} from '../../_gqlTypes/GET_LIBRARIES';

interface ILibrariesListProps {
    libraries: GET_LIBRARIES_libraries[] | null;
    t: TranslationFunction;
    i18n: i18n;
    onRowClick: (library: GET_LIBRARIES_libraries) => void;
}

function LibrariesList({libraries, onRowClick, t, i18n: i18next}: ILibrariesListProps): JSX.Element {
    return (
        <React.Fragment>
            <Table selectable striped>
                <Table.Header>
                    <Table.Row>
                        <Table.HeaderCell width={1} />
                        <Table.HeaderCell width={7}>{t('libraries.label')}</Table.HeaderCell>
                        <Table.HeaderCell width={6}>{t('libraries.ID')}</Table.HeaderCell>
                        <Table.HeaderCell width={1}>{t('libraries.isSystem')}</Table.HeaderCell>
                        <Table.HeaderCell width={1} />
                    </Table.Row>
                </Table.Header>
                <Table.Body>
                    {libraries &&
                        libraries.map(l => {
                            const libLabel = localizedLabel(l.label, i18next);
                            const onClick = () => onRowClick(l);
                            return (
                                <Table.Row key={l.id} onClick={onClick}>
                                    <Table.Cell>
                                        <Icon name="book" size="large" />
                                    </Table.Cell>
                                    <Table.Cell>{libLabel}</Table.Cell>
                                    <Table.Cell>{l.id}</Table.Cell>
                                    <Table.Cell>
                                        <Checkbox readOnly checked={!!l.system} />
                                    </Table.Cell>
                                    <Table.Cell>
                                        <DeleteLibrary library={l} />
                                    </Table.Cell>
                                </Table.Row>
                            );
                        })}
                </Table.Body>
            </Table>
        </React.Fragment>
    );
}

export default translate()(LibrariesList);
