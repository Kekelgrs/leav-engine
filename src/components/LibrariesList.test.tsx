import {shallow} from 'enzyme';
import * as React from 'react';
import {BrowserRouter as Router} from 'react-router-dom';
import {create} from 'react-test-renderer';
import {List} from 'semantic-ui-react';
import LibrariesList from './LibrariesList';

describe('LibrariesList', () => {
    test('Render list', async () => {
        const libraries = [
            {id: 'test', label: {fr: 'Test', en: null}},
            {id: 'test2', label: {fr: null, en: 'Test 2'}},
            {id: 'test3', label: null}
        ];

        const comp = <LibrariesList libraries={libraries} />;
        const shallowComp = shallow(comp);

        const listItems = shallowComp.find(List.Item);

        expect(listItems.length).toBe(3);
        expect(
            listItems
                .at(0)
                .find(List.Header)
                .first()
                .props().children
        ).toEqual('Test');

        expect(
            listItems
                .at(1)
                .find(List.Header)
                .first()
                .props().children
        ).toEqual('Test 2');

        expect(
            listItems
                .at(2)
                .find(List.Header)
                .first()
                .props().children
        ).toEqual('test3');

        const renderedComp = create(<Router>{comp}</Router>).toJSON();
        expect(renderedComp).toMatchSnapshot();
    });
});
