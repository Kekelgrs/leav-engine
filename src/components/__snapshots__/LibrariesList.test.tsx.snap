// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibrariesList Render list 1`] = `
<div
  className="ui divided relaxed list"
  role="list"
>
  <a
    className="item"
    data-value={undefined}
    href="/libraries"
    onClick={[Function]}
    role="listitem"
  >
    <i
      aria-hidden="true"
      className="book large icon"
    />
    <div
      className="content"
    >
      <div
        className="header"
        size="huge"
      >
        Test
      </div>
      <div
        className="description"
      >
        test
      </div>
    </div>
  </a>
  <a
    className="item"
    data-value={undefined}
    href="/libraries"
    onClick={[Function]}
    role="listitem"
  >
    <i
      aria-hidden="true"
      className="book large icon"
    />
    <div
      className="content"
    >
      <div
        className="header"
        size="huge"
      >
        Test 2
      </div>
      <div
        className="description"
      >
        test2
      </div>
    </div>
  </a>
  <a
    className="item"
    data-value={undefined}
    href="/libraries"
    onClick={[Function]}
    role="listitem"
  >
    <i
      aria-hidden="true"
      className="book large icon"
    />
    <div
      className="content"
    >
      <div
        className="header"
        size="huge"
      >
        test3
      </div>
      <div
        className="description"
      >
        test3
      </div>
    </div>
  </a>
</div>
`;
